samples/nrf9160/http_application_update/src/main.c

```diff
@@ -0,0 +1,271 @@
+static volatile bool	start_dfu;
+static struct		device *gpiob;
+static struct		gpio_callback gpio_cb;
+static u32_t		flash_address;
+
+static int app_download_client_event_handler(struct download_client *const dfu,
+			enum download_client_evt event, u32_t status);
+
+
+static struct download_client dfu = {
+	.host = CONFIG_DOWNLOAD_HOST,
+	.resource = CONFIG_DOWNLOAD_FILE,
+	.callback = app_download_client_event_handler
+};
+
+
+/**@brief Recoverable BSD library error. */
+void bsd_recoverable_error_handler(uint32_t err)
+{
+	printk("bsdlib recoverable error: %u\n", err);
+}
+
+/**@brief Irrecoverable BSD library error. */
+void bsd_irrecoverable_error_handler(uint32_t err)
+{
+	printk("bsdlib irrecoverable error: %u\n", err);
+
+	__ASSERT_NO_MSG(false);
+}
+
+
+/**@brief Initialize application. */
+static int app_dfu_init(void)
+{
+	int i;
+
+	flash_address = DT_FLASH_AREA_IMAGE_1_OFFSET;
+	for (i = 0; i < FLASH_PAGE_MAX_CNT; i++) {
+		is_flash_page_erased[i] = false;
+	}
+	flash_dev = device_get_binding(DT_FLASH_DEV_NAME);
+	if (flash_dev == 0) {
+		printk("Nordic nRF flash driver was not found!\n");
+		return 1;
+	}
+
+	int retval = download_client_init(&dfu);
+
+	if (retval != 0) {
+		printk("download_client_init() failed, err %d", retval);
+		return 1;
+	}
+	return 0;
+}
+
+
+/**@brief Start transfer of the file. */
+static int app_dfu_transfer_start(void)
+{
+	int retval = download_client_connect(&dfu);
+
+	if (retval != 0) {
+		printk("download_client_connect() failed, err %d",
+			retval);
+		return 1;
+	}
+
+	retval = download_client_start(&dfu);
+	if (retval != 0) {
+		printk("download_client_start() failed, err %d",
+			retval);
+		return 1;
+	}
+	return 0;
+}
+
+
+static int app_download_client_event_handler(
+	struct download_client *const dfu,
+	enum download_client_evt event, u32_t error)
+{
+	int err;
+
+	switch (event) {
+	case DOWNLOAD_CLIENT_EVT_DOWNLOAD_FRAG: {
+		struct flash_pages_info info;
+
+		if (dfu->object_size > DT_FLASH_AREA_IMAGE_1_SIZE) {
+			printk("Requested file too big to fit in flash\n");
+			return 1;
+		}
+		err = flash_get_page_info_by_offs(flash_dev, flash_address,
+			&info);
+		if (err != 0) {
+			printk("flash_get_page_info_by_offs returned error %d\n",
+				err);
+			return 1;
+		}
+		flash_write_protection_set(flash_dev, false);
+		if (!is_flash_page_erased[info.index]) {
+			err = flash_erase(flash_dev, flash_address, info.size);
+			if (err != 0) {
+				printk("flash_erase returned error %d at address %08x\n",
+					err, flash_address);
+				return 1;
+			}
+			is_flash_page_erased[info.index] = true;
+		}
+		err = flash_write(flash_dev, flash_address, dfu->fragment,
+							dfu->fragment_size);
+		flash_write_protection_set(flash_dev, true);
+		if (err != 0) {
+			printk("Flash write error %d at address %08x\n",
+				err, flash_address);
+			return 1;
+		}
+		flash_address += dfu->fragment_size;
+		break;
+	}
+
+	case DOWNLOAD_CLIENT_EVT_DOWNLOAD_DONE:
```

Comment:
Need to check whether the last image sector was erased and erase it if was not erased. Image status land in this sector so it must be erased during successful image collection.

Where boot_request_upgrade() is called?

Reply:
I have now added a call to `boot_request_upgrade(0)` in DOWNLOAD_CLIENT_EVT_DOWNLOAD_DONE. The last page in primary slot is erased if needed. Since the page erase code is repeated two times I refactored that out to a common function `flash_page_erase_if_needed`

Reply:
LGTM